CIPs/cip-granda_mento.md

```diff
@@ -0,0 +1,187 @@
+Currently, `StableToken.sol` only allows its `Exchange.sol` or `Validators.sol` to [mint](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L225) new stable tokens. Similarly, only its `Exchange.sol` can [burn](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L275) stable tokens.
+
+The following modifications will be made:
+1. **Modification to [mint()](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L224)**
+   * Allow `msg.sender` to be `GrandaMento` in addition to the existing permitted senders, Exchange and `Validators.sol`.
+2. **Modification to [burn()](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L273)**
+   * Allow `msg.sender` to be `GrandaMento` in addition to the existing permitted sender, the Exchange.
+
+#### New contracts:
+
+##### `GrandaMento.sol`
+
+A new contract, `GrandaMento.sol`, is created, added to the Registry with the identifier `GrandaMento`, and owned by Governance. `GrandaMento` is a singleton contract that is capable of facilitating exchanges for multiple stable tokens-- that is, with the approval of Governance, `GrandaMento` could support `CELO <-> cUSD`, `CELO <-> cEUR`, and future `CELO <-> cXXX` exchanges. `GrandaMento` must be a spender of `Reserve.sol`, and supported stable tokens must allow `GrandaMento` to mint/burn them.
+
+The contract has the following configurable parameters:
+
+1. **`address public approver;`** - Set by Governance. Intended to be a multisig and has the authority to approve exchanges. The signers for the multisig are at the discretion of Governance, but it would comprise of community members or members on behalf of organizations that are aligned with the Celo network/ecosystem.
+2. **`uint256 public exchangeWaitPeriodSeconds;`** - Set by Governance. The minimum amount of time in seconds that must elapse between a proposed exchange being approved and when the exchange can be executed. Should give sufficient time for Governance to veto an approved exchange.
+3. **`FixidityLib.Fraction public spread;`** - Set by Governance. The percent fee imposed upon an exchange execution.
+4. **`mapping(address => ExchangeLimits) stableTokenExchangeLimits`** - Set by Governance via `setStableTokenExchangeLimits`. A mapping of stable token address to a struct containing two `uint256`s, `minExchangeAmount` and `maxExchangeAmount`.
+
+The contract has the following functions:
+
+1. **`function proposeExchange(address stableToken, uint256 sellAmount, bool sellCelo) external returns (uint256)`** - Called by an exchange proposer to propose an exchange.
+   * Callable by anyone.
+   * If `sellCelo` is true, CELO is the asset being sold and `stableToken` is the asset being bought. If `sellCelo` is false, `stableToken` is the asset being sold and CELO is the asset being bought.
```

Comment:
I know we use this `sellCelo` switch elsewhere but I have always found this confusing
Can we just expose 2 endpoints? Can you remind me of the rationale here.

Reply:
I wasn't involved with the rationale in `Exchange.sol` but I feel like the 2 endpoint approach would end up looking like (excuse the bad fn names):
```
function proposeExchange(address stableToken, uint256 sellAmount, bool sellCelo) internal returns (uint256) {
  // ...
}

function proposeCeloSell(address stableToken, uint256 sellAmount) external returns (uint256) {
  return proposeExchange(stableToken, sellAmount, true);
}

function proposeStableTokenSell(address stableToken, uint256 sellAmount) external returns (uint256) {
  return proposeExchange(stableToken, sellAmount, false);
}
```

And imo there isn't much to gain from having those functions on the contract level. Though I usually think of contracts as something that's ok to have a slightly rough interface because they're usually going to be called by another piece of code (like contractkit or the cli) that can abstract away the rough interface and give a better UX. Does CAP think of contracts differently?

Reply:
For the core contracts, usage by another contract is probably the most pertinent. Much of this proposal seems motivated by mitigating misuse of the exchange proposals. User understanding of verified interfaces is one of the last lines of defense against malicious application signatures, so I generally prefer a more specified interface where possible. For example, the user probably knows nothing about the address of `address stableToken`, but could more reasonably sign a `proposeCELOToCUSDExchange(amount)`.

Reply:
I have advocated for registry ids in favor of addresses in the past and will continue to do so for this reason, among others.

Reply:
For posterity, ended up using the registry ID of the stable token as a parameter. Also moved to using the registry ID as the key for the exchange limits of a stable token.